### ISSUE TITLE
GsonMessageBodyProvider not handling String.class passed into writeTo

### ISSUE THREAD

**Issue by dgrant** [immutables/immutables#562]
In GsonMessageBodyProvider, I see this in readFrom():

```
      if (ex.getCause() instanceof RuntimeException) {
        String json = gson.toJson(new Error(ex.getCause().getMessage()));
        throw new WebApplicationException(
            Response.status(Status.BAD_REQUEST)
                .type(mediaType)
                .entity(json)
                .build());
      }
```

The problem here is that Restlet is handling the incoming request, it asks GsonMesssageBodyProvider to convert the incoming json to an object, that fails, and throws a NullPointerException (because one of the fields was null in the json). It then gets into the code above where it creates a response with {"error": "exception message"} in the body.

Then GsonMessageProvider is called again to convert this response object to a json string. But it's already a json string and it has a type token of null so this causes an exception in write() because type is null.

Anyways, what should happen is that this:

```
        String json = gson.toJson(new Error(ex.getCause().getMessage()));
        throw new WebApplicationException(
            Response.status(Status.BAD_REQUEST)
                .type(mediaType)
                .entity(json)
                .build());
```

should just be a re-throw of the RuntimeException. A RuntimeException, for us, will cause a 500 to be returned to our frontend and the exception to be logged.

**Comment by elucash** [comment]
Seems that Restlet violates specification of `WebApplicationException`:

``` 
 * This exception may be thrown by a resource method, provider or
 * {@link javax.ws.rs.core.StreamingOutput} implementation if a specific
 * HTTP error response needs to be produced. Only effective if thrown prior to
 * the response being committed.
```
It is a good idea to post Restlet bug. What I can propose on the side of Immutables is a way to install custom error handler to do whatever needed. But changing the behavior to throw RuntimeException doesn't sound right.

**Comment by elucash** [comment]
Just to note: I've inspected a couple message body providers from Jersey implementation, like JAXB one, etc, they tend to throw `WebApplicationException(400)`  too.

**Comment by dgrant** [comment]
@elucash Thanks for your response. I really don't know tons about this stuff which is why I filed a bug here instead of either reporting a bug to Restlet or submitting a PR to Restlet or Immutables. So just to get that context out of the way...

I will step through with a debugger after this and focus more on Restlet code rather than Immutables.

Can you articulate what you think exactly that Restlet is doing wrong?

I did test this a while ago with your unit tests (where you use Jersey) and it seemed to do what I expected, there was not call GsonMessageBodyProvider for the response-side... once the GsonMessageBodyProvider encountered an NPE on the request-side, it errored back to the client. Restlet, on the other hand, seems to still call the GsonMessageBodyProvider. Is that the issue? That it shouldn't do anything on the response but just return 400 straight away?

**Comment by dgrant** [comment]
Ok, I think I found the area in Restlet where things are weird/wrong

JaxRsRestlet.convertToRepresentation is called to convert the response that Immutuables creates "{"error", "missingFieldNameHere"}". It calls JaxRsProviders.writerSubSet to get the writers that can handle the type String.class. Apparently GsonMessageBodyProvider passes that test... and it is included in the list of writers that can handle String.class. It seems that in AbstractProviderWrapper, it determines that GsonMessageBodyProvider handles any Object.class, because it implements MessageBodyReader<Object>, MessageBodyWriter<Object>...

So it has a list of possible writers now, including GsonMessageBodyProvider. It now tries to determine the best one. It narrows it down to GsonMessageBodyProvider, and a few restlet ones: ConverterProvider, StringProvider, and JsonProvider. It ends up taking GsonMessageBodyProvider out of those 4 candidates simply because it is first in the list.

Hmm, now I kind of get lost at this point....

I've put a breakpoint in GsonMessageBodyProvider and it Restlet is definitely calling it to try to convert the String object "{"error", "missingFieldNameHere"}" to a String. I noticed that in GsonMessageBodyProvider, isWriteable is being called with type==String.class... isWriteable seems to ignore what the type is though. Should GsonMessageBodyProvider return false if there is a string there? I mean there is no way a String can be serialized to json right? It doesn't make any sense?

I'm looking at a JacksonJsonProvider, which is another implementation of MessageBodyWriter, and it does this inside it's isWriteable:

```
if(_untouchables.contains(new ClassKey(type))) {
            return false;
}

// where _untouchables is:
        _untouchables.add(new ClassKey(InputStream.class));
        _untouchables.add(new ClassKey(Reader.class));
        _untouchables.add(new ClassKey(OutputStream.class));
        _untouchables.add(new ClassKey(Writer.class));
        _untouchables.add(new ClassKey(byte[].class));
        _untouchables.add(new ClassKey(char[].class));
        _untouchables.add(new ClassKey(String.class));
        _untouchables.add(new ClassKey(StreamingOutput.class));
        _untouchables.add(new ClassKey(Response.class));
```

**Comment by dgrant** [comment]
Re-titled this issue now that I understand better what's going on

**Comment by elucash** [comment]
My earlier point was about that `WebApplicationException` should be just used to produce a response and should not be handled in "custom" manner, but probably I totally misunderstood the actual issue and intricacies of the JAX-RS implementations.
I've made some revisions to the implementation of `GsonMessageBodyProvider` to be able to attach custom error handler as well as "safer" behavior for certain types. I would encourage you to review and try it to see if this can solve the problem

**Comment by dgrant** [comment]
@elucash Sorry yeah I didn't understand the problem initially either... I think the WebApplicationException is just fine for now... I think your fix will get rid of the issue we're seeing

**Comment by dgrant** [comment]
So we had an immediate need yesterday for this fix, as one of our developers actually wanted to shove a String into the body of a response (and not use an immutable object in this case). Tried build on latest master and it seems to work great so far.

**Comment by elucash** [comment]
good news! hope to release soon

**Comment by elucash** [comment]
released in 2.4.4. thank you!

**Comment by dgrant** [comment]
Thank you!
